cypress/e2e/swap.test.ts

```diff
@@ -194,4 +197,56 @@ describe('Swap', () => {
     cy.get('#swap-currency-input .token-amount-input').should('have.value', '')
     cy.get('#swap-currency-output .token-amount-input').should('not.equal', '')
   })
+
+  it('should render an error for slippage failure', () => {
+    const SLIPPAGE = 0.01
+    const ETH_GAS_BUFFER = '10'
+    cy.visit('/swap', { ethereum: 'hardhat' })
+      .hardhat()
+      .then((hardhat) => {
+        hardhat
+          .reset()
+          .then(() => hardhat.provider.send('evm_setAutomine', [false]))
+          .then(() => hardhat.provider.getBalance(hardhat.wallet.address))
+          .then((initialBalance) => {
+            const swapInputAmount = formatEther(initialBalance.sub(parseEther(ETH_GAS_BUFFER)))
+            const send = cy.stub(hardhat.provider, 'send')
+            send.withArgs('eth_estimateGas').resolves(BigNumber.from(2_000_000))
+            send.callThrough()
+
+            // Set slippage to a very low value.
+            cy.get(getTestSelector('open-settings-dialog-button')).click()
+            cy.get(getTestSelector('slippage-input')).clear().type(SLIPPAGE.toString())
+            cy.get('body').click('topRight')
+            cy.get(getTestSelector('slippage-input')).should('not.exist')
+
+            // Select USDC as output token from the common bases menu.
+            cy.get('#swap-currency-output .open-currency-select-button').click()
+            cy.contains('USDC').click()
+
+            // Set the input amount to the max amount minus the gas buffer.
+            cy.get('#swap-currency-input .token-amount-input').clear().type(swapInputAmount)
+            cy.get('#swap-currency-output .token-amount-input').should('not.equal', '')
+            cy.get('#swap-button').click()
+            cy.get('#confirm-swap-or-send').click()
+            cy.get(getTestSelector('dismiss-tx-confirmation')).click()
+
+            cy.then(() => hardhat.provider.send('hardhat_mine', ['0x2', '0xc'])).then(() => {
+              // Check for a failed transaction notification.
+              cy.contains('Swap failed').should('exist')
+
+              // Assert that the balance is unchanged. (aside from gas costs)
+              cy.then(() => hardhat.provider.getBalance(hardhat.wallet.address)).then((finalBalance) => {
+                expect(initialBalance.gt(finalBalance)).to.be.true
+                expect(finalBalance.gt(ETH_GAS_BUFFER)).to.be.true
```

Comment:
Yes, but this assertion gets us the same degree of safety with substantially less work. Since we're only working with a single transaction we don't need to know *exactly* how much gas was consumed and how it was priced at the fork point.